src/emulator.rs

```diff
@@ -79,11 +79,35 @@ impl Cpu<u64> {
     // #### Control
 
     // `beq rs1,rs2,imm`: `if (rs1 == rs2) { pc = pc + imm } else { pc = pc + 4 }` with `-2^12 <= imm < 2^12` and `imm % 2 == 0`
-    // TODO
+    pub fn beq(&mut self, rs1: usize, rs2: usize, imm: i64) {
+        if self.regs[rs1] == self.regs[rs2] {
+            assert!(imm < 2 << 12 && imm >= -(2 << 12));
+            assert_eq!(imm % 2, 0);
+            assert_eq!(imm % 4, 0, "instruction-address-misaligned");
+            let pc = self.pc as i128;
+            self.pc = (pc + imm as i128) as u64;
```

Comment:
This comment also applies to the other instructions:  I think there's no need to convert to `i128`. RV64 works with 64 bit registers, so the address space can't be higher than 64 bits. In theory some edge case would overflow the pc, and that can already be calculated with `i64` arithmetic.  So I think this can look like this just fine:

```suggestion
            let pc = self.pc as i64;
            self.pc = (pc + imm as i64) as u64;
```
I think it's important to make this change so that the emulator and the circuit behave the same: in the circuit/lookups doing addition of 128 bits needs different logic than addition of 64 bits.

Reply:
That makes sense. Exactly I thought that `pc + imm` could be overflow but it is `i64` arithmetic.